startout/starterfile.py

```diff
@@ -0,0 +1,401 @@
+
+                destroyed_tools = []
+                for tool in [tool for tool in self.tools if tool.name in successful_tools]:
+                    if not tool.destroy():
+                        # TODO handle failure to destroy better
+                        print(f"FATAL: Failed to destroy tool \"{tool.name}\"", file=sys.stderr)
+                        print(f".. Only destroyed these tools: {destroyed_tools}", file=sys.stderr)
+                        sys.exit(1)
+                    else:
+                        destroyed_tools.append(tool.name)
+
+            # If any tools failed
+            return False
+
+        # If all tools were successfully installed or were already installed
+        return True
+
+    def install_modules(self, teardown_on_failure=True, fail_early=True):
+        """
+        Install modules layer by layer so that their dependencies are all met before being installed.
+
+        :param teardown_on_failure: If True, rollback other tools if any module installation fails.
+        :type teardown_on_failure: bool
+        :param fail_early: If True, function will return false as soon as a tool fails to initialize.
+        :type fail_early: bool
+        :return: True if all modules are installed successfully, False otherwise.
+        :rtype: bool
+        """
+
+        # Early exit if there are no modules to install.
+        if len(self.modules) == 0:
+            print("Nothing to do.")
+            return False
+
+        print("Installing modules...")
+
+        failed_modules = []
+        successful_modules = []
+        for layer in self.module_dependencies:
+            # Install modules layer by layer so that their dependencies are all met before being installed
+            early_exit = False
+
+            for module in (module for module in self.modules if module.name in layer):
+                # Initialize this module, adding it to the list of failures if it cannot be initialized
+                if not module.initialize():
+                    failed_modules.append(module.name)
+                    if fail_early:
+                        early_exit = True
+                else:
+                    successful_modules.append(module.name)
+
+            if early_exit:
+                break
+
+        # Report any failed modules
+        if len(failed_modules) > 0:
+            print("Failed modules:", failed_modules, file=sys.stderr)
+
+            # Destroy modules which were successfully installed if specified
+            if teardown_on_failure:
+                print("Rolling back successfully installed modules:",
+                      successful_modules,
+                      file=sys.stderr)
+
+                destroyed_modules = []
+                for module in [module for module in self.modules if module.name in successful_modules]:
+                    if not module.destroy():
+                        # TODO handle failure to destroy better
```

Comment:
Found this TODO, unsure if we want to merge it down or handle it first

Reply:
I think that later on down the road we can define authors of modules/help docs that could be useful in these scenarios, but if a module/tool fails to destroy then it's likely outside of the scope of the CLI to resolve. Thank you for finding this! Once we integrate todoon it'll generate an issue for us lol